src/mpid/ch4/src/ch4r_rma.h

```diff
@@ -501,14 +536,141 @@ MPL_STATIC_INLINE_PREFIX int MPIDI_CH4I_do_accumulate(const void *origin_addr,
     if (mpi_errno)
         MPIR_ERR_POP(mpi_errno);
 
-    MPIR_FUNC_VERBOSE_EXIT(MPID_STATE_CH4U_DO_ACCUMULATE);
   fn_exit:
+    *sreq_ptr = sreq;
+
+    MPIR_FUNC_VERBOSE_EXIT(MPID_STATE_CH4U_DO_GET_ACCUMULATE);
+    return mpi_errno;
+  fn_fail:
+    goto fn_exit;
+}
+
+#undef FUNCNAME
+#define FUNCNAME MPIDI_CH4U_mpi_put
+#undef FCNAME
+#define FCNAME MPL_QUOTE(FUNCNAME)
+MPL_STATIC_INLINE_PREFIX int MPIDI_CH4U_mpi_put(const void *origin_addr,
+                                                int origin_count,
+                                                MPI_Datatype origin_datatype,
+                                                int target_rank,
+                                                MPI_Aint target_disp,
+                                                int target_count, MPI_Datatype target_datatype,
+                                                MPIR_Win * win)
+{
+    int mpi_errno = MPI_SUCCESS;
+    MPIR_Request *sreq = NULL;
+    MPIR_FUNC_VERBOSE_STATE_DECL(MPID_STATE_CH4U_PUT);
+    MPIR_FUNC_VERBOSE_ENTER(MPID_STATE_CH4U_PUT);
+
+    mpi_errno = do_put(origin_addr, origin_count, origin_datatype,
+                       target_rank, target_disp, target_count, target_datatype, win, &sreq);
+    if (mpi_errno)
+        MPIR_ERR_POP(mpi_errno);
```

Comment:
I'm not sure what the point of the `do_put` function is.  It looks like that's exactly the same as the `rput` function.  Why not simply add all that content in `rput`, and have `put` call `rput` instead?  Same comment for the remaining ops as well.